db.go

```diff
@@ -309,17 +314,82 @@ func (db *DB) GetMany(ro *ReadOptions, keys [][]byte) ([][]byte, []error) {
 		if valueLens[i] > 0 {
 			values[i] = C.GoBytes(unsafe.Pointer(uintptr(unsafe.Pointer(cPackedVals))+uintptr(valOffset)), C.int(valueLens[i]))
 			valOffset += int(valueLens[i])
-		} else if (errs == nil || errs[i] == nil) && valueLens[i] == 0 {
+		} else if (err == nil || err.(*MultiKeyError).errAt(i) == nil) && valueLens[i] == 0 {
 			// No error and no value, it means keys[i] is found but with empty value
 			values[i] = []byte{}
 		}
 		// else we have either:
-		// 1. (errs != nil && errs[i] == nil) && valueLens[i] < 0 indicating keys[i] is not found
-		// 2. (errs != nil && errs[i] != nil) indicating an error looking up keys[i].
+		// 1. (err != nil && multiKeyErr.errAt(i) == nil) && valueLens[i] < 0 indicating keys[i] is not found
+		// 2. (err != nil && multiKeyErr.errAt(i) != nil) indicating an error looking up keys[i].
 		// In both cases values[i] defaults to nil
 	}
 
-	return values, errs
+	return values, err
+}
+
+// MultiKeyError encapsulates multiple errors encountered in a Get/Put-Many() call,
+// behind the errors.error interface. Caller may cast the generic error object to
+// a MultiKeyError and call Errors() and/or FailedKeyIndexes() to get additional details.
+type MultiKeyError struct {
```

Comment:
nit: can we separate in another file ?

Reply:
sure. I was thinking it may make merging upstream more of a hassle, but probably doesn't matter